packages/mui-material/src/IconButton/IconButton.js

```diff
@@ -113,7 +113,7 @@ const IconButtonRoot = styled(ButtonBase, {
       ...Object.entries(theme.palette)
         .filter(createSimplePaletteValueFilter()) // check all the used fields in the style below
         .map(([color]) => ({
-          props: { color, disableRipple: false },
+          props: (props) => props.color === color && !props.disableRipple,
```

Comment:
@siriwatknp If I remember correctly there's a problem with accessing color like this, isn't there? How might we implement this one?

Reply:
Ah, you are right. This won't work with Pigment CSS.

Reply:
Here is my proposal using CSS variable: https://github.com/sai6855/material-ui/pull/1.

We keep the previous logic and separate the `disableRipple` from the variants.

Reply:
> Here is my proposal using CSS variable: [sai6855#1](https://github.com/sai6855/material-ui/pull/1).
> 
> We keep the previous logic and separate the `disableRipple` from the variants.

Sure, @DiegoAndai . I'm not really familiar with the changes, but if you give the go-ahead, I can merge the PR.

Reply:
Thanks @siriwatknp! This solution makes sense 👌🏼 

@sai6855 may I ask you to update this PR with that solution?

Reply:
@DiegoAndai I've merged the PR https://github.com/sai6855/material-ui/pull/1, but looks like few tests are failing https://circleci.com/gh/mui/material-ui/748097. Please feel free to make changes 

cc @siriwatknp

Reply:
@sai6855 I'll take a look, I have an idea to improve the tests

Reply:
My idea didn't work 😅 

The problem is that with this change, we're using a CSS variable that is not read properly when asserting with `toHaveComputedStyle`. So as an example if before we had:

```js
// style:
const Element = styled('div')({
  backgroundColor: 'red'
})

// assert
expect(element).toHaveComputedStyle({ backgroundColor: 'rgb(255, 0, 0)' }); // passes ✅ 
```

This updated code isn't passing:

```js
// style:
const Element = styled('div')({
  '--background-color-variable': 'red',
  backgroundColor: 'var(--background-color-variable)',
})

// assert
expect(element).toHaveComputedStyle({ backgroundColor: 'rgb(255, 0, 0)' }); // fails 🔴 
```

You can see this in the video:

https://github.com/user-attachments/assets/a990b2f5-38b1-4714-b3c2-7530d44f0c46

-------

Unfortunately, the test is taking way more time than the actual fix 😂 I'll take another look on Monday to see if I can come up with alternatives.

cc: @Janpot this is what I was commenting you about JSDOM and CSS Variables.

Reply:
> cc: @Janpot this is what I was commenting you about JSDOM and CSS Variables.

Looks like we're hitting https://github.com/jsdom/jsdom/issues/3563 here. I guess we'll have to use a real browser for some of these tests.

Reply:
>  I guess we'll have to use a real browser for some of these tests.

I'm also getting the same behavior when running these with Karma, though 😓:

<img width="700" alt="Screenshot 2024-09-16 at 16 52 52" src="https://github.com/user-attachments/assets/9addae2a-4ce5-47f6-bbd6-0990dc72a615">

Which is unexpected. Could something be wrong in our/my setup? Or am I missing something?

Reply:
@DiegoAndai That's odd, as a quick test I added 

```tsx
  it.only('quick test', () => {
    const Element = styled('div')({
      '--background-color-variable': 'red',
      backgroundColor: 'var(--background-color-variable)',
    });

    const { container, debug } = render(<Element />);
    debug(container.firstChild);
    console.log(window.getComputedStyle(container.firstChild).backgroundColor);
    expect(container.firstChild).toHaveComputedStyle({ backgroundColor: 'rgb(255, 0, 0)' });
  });
```

to `Accordion.test.js` and ran `pnpm test:karma`. This seems to work as expected.

I'm seeing a suspicious log message in your screenshot, are you sure JSDOM is not just running in the browser?

Reply:
> I'm seeing a suspicious log message in your screenshot, are you sure JSDOM is not just running in the browser?

I don't understand the question 😅

Reply:
nevermind, I thought the "Styles in JSDOM e.g. from..." message was logged by JSDOM, but it comes from our `assertMatchingStyles`

Reply:
@Janpot I can confirm that [your quick test](https://github.com/mui/material-ui/pull/43714#discussion_r1764751568) works, as well as CSS variables in the IconButton component when running with Karma.

I discovered that the issue I'm facing, and why the test is not working, is that for some reason I can't get the hover style to be applied. I assume this is the reason the existing test is skipped in Karma: https://github.com/mui/material-ui/blob/master/packages/mui-material/src/IconButton/IconButton.test.js#L146

I've tried `fireEvent.mouseMove`, `fireEvent.mouseOver`, `fireEvent.mouseEnter`, and `userEvent.hover`, and none are working.

Reply:
But we are currently using `fireEvent` in JSDOM to test the hover style: https://github.com/mui/material-ui/blob/master/packages/mui-material/src/IconButton/IconButton.test.js#L152

This is the only test we have that uses this, though, and it seems like:

- JSDOM handles hover but not CSS variables
- Browser tests handle CSS variables but not hover 
😅 

At this point, I'm considering removing the test and opening an issue to properly implement hover tests in the future. What do you think?

Reply:
We're working on a migration to `vitest` and in their browser mode we'll have access to real user events https://vitest.dev/guide/browser/context.html. Doesn't look like karma has an equivalent. Maybe just skip the test for now?

Reply:
👍🏼 agree